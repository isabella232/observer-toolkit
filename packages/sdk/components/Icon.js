import React from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import icons from '../theme/icons'

const Container = styled.span`
  user-select: none; /* Stops nearby text being selected if icon clicked quickly */
  display: inline-block;
  vertical-align: middle;
  width: ${({ size }) => size};
  height: ${({ size }) => size};

  ${({ theme, active }) => active && `color: ${theme.color('highlight')};`}

  ${({ theme, disabled }) =>
    disabled &&
    `
      color: ${theme.color('text', 3)};
  `}
  ${({ onClick, theme, disabled }) =>
    onClick &&
    !disabled &&
    `
    cursor: pointer;
    &:hover {
      backgroundColor: ${theme.color('background', 1, 0.3)};
    }
  `}
`

const Empty = styled.span`
  display: inline-block;
`

function Icon({
  type,
  onClick,
  active,
  disabled,
  size = '2em',
  override = {},
  ...props
}) {
  const IconSvg = type ? icons[type] : Empty
  if (!IconSvg) throw new Error(`No icon found named "${type}"`)

  return (
    <Container
      onClick={disabled ? null : onClick}
      active={active}
      disabled={disabled}
      size={size}
      as={override.Container}
      {...props}
    >
      <IconSvg as={override.IconSvg} />
    </Container>
  )
}

Icon.propTypes = {
  type: T.string,
  onClick: T.func,
  active: T.bool,
  disabled: T.bool,
  size: T.string,
  override: T.object,
}

export default Icon
