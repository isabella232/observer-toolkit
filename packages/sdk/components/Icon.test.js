import React from 'react'
import { fireEvent } from '@testing-library/react'
import { renderWithTheme } from '@libp2p-observer/testing'
import Icon from './Icon'

describe('Icon', () => {
  it('throws if trying to render an invalid icon', () => {
    expect(() => renderWithTheme(<Icon type="invalid icon type" />)).toThrow(
      new Error('No icon found named "invalid icon type"')
    )
  })

  it("renders 'asc' icon as expected", () => {
    const { asFragment } = renderWithTheme(<Icon type="asc" disabled={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('acts as button calling onClick if clicked', () => {
    const mockFn = jest.fn()

    const { getByRole } = renderWithTheme(
      <Icon type="asc" disabled={false} onClick={mockFn} />
    )
    fireEvent.click(getByRole('button'))
    expect(mockFn).toBeCalled()
  })

  it("doesn't act as button or call onClick if disabled", () => {
    const mockFn = jest.fn()

    const { getByRole, queryByRole } = renderWithTheme(
      <Icon type="asc" disabled={true} onClick={mockFn} />
    )
    expect(queryByRole('button')).toBeNull()

    fireEvent.click(getByRole('img'))
    expect(mockFn).not.toBeCalled()
  })
})
