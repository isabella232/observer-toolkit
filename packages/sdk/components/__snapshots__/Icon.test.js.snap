// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders 'asc' icon as expected 1`] = `
.c0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 20px;
  height: 20px;
  border-radius: 20px;
  color: rgb(255,255,255);
}

<span
  className="c0"
  disabled={true}
  onClick={null}
  size={20}
>
  <svg
    fill="currentColor"
    strokeLinejoin="round"
    strokeMiterlimit={1.414}
    viewBox="16 16 70 70"
  >
    <path
      d="M51.494 35.157l4.41 4.409a1.5 1.5 0 102.121-2.121l-6.965-6.964a1.5 1.5 0 00-2.12 0l-6.957 6.956a1.5 1.5 0 102.121 2.121l4.39-4.389v33.367a1.503 1.503 0 001.5 1.5 1.503 1.503 0 001.5-1.5V35.157z"
    />
  </svg>
</span>
`;
